src/index/sparse/sparse_inverted_index.h

```diff
@@ -0,0 +1,449 @@
+        } else {
+            search_wand(q_vec, heap, bitset);
+        }
+
+        // no refinement needed
+        if (refine_factor == 1) {
+            collect_result(heap, distances, labels);
+        } else {
+            std::unordered_map<IndicesT, T> q_map;
+            for (size_t i = 0; i < len; ++i) {
+                q_map[indices[i]] = data[i];
+            }
+            refine_and_collect(q_map, heap, k, distances, labels);
+        }
+    }
+
+    [[nodiscard]] size_t
+    size() const {
+        size_t res = 0;
+        res += sizeof(*this);
+        res += sizeof(T) * data_.capacity();
+        res += sizeof(IndicesT) * indices_.capacity();
+        res += sizeof(IndPtrT) * indptr_.capacity();
+        res += sizeof(std::vector<Neighbor<T>>) * inverted_lut_.capacity();
+        for (auto& lut : inverted_lut_) {
+            res += sizeof(Neighbor<T>) * lut.capacity();
+        }
+        if (use_wand_) {
+            res += sizeof(T) * max_in_dim_.capacity();
+        }
+        return res;
+    }
+
+    [[nodiscard]] size_t
+    n_rows() const {
+        return n_rows_;
+    }
+
+    [[nodiscard]] size_t
+    n_cols() const {
+        return n_cols_;
+    }
+
+ private:
+    [[nodiscard]] float
+    dot_product(std::unordered_map<IndicesT, T> q_map, table_t u) const {
+        float res = 0.0f;
+        for (IndPtrT i = indptr_[u]; i < indptr_[u + 1]; ++i) {
+            auto idx = indices_[i];
+            float val = float(data_[i]);
+            auto it = q_map.find(idx);
+            if (it != q_map.end()) {
+                res += val * it->second;
+            }
+        }
+        return res;
+    }
+
+    // find the top-k candidates using brute force search, k as specified by the capacity of the heap.
+    void
+    search_brute_force(std::vector<std::pair<IndicesT, T>>& q_vec, MinMaxHeap<T>& heap,
+                       const BitsetView& bitset) const {
+        std::vector<float> scores(n_rows_, 0.0f);
+        for (auto [i, v] : q_vec) {
+            for (size_t j = 0; j < inverted_lut_[i].size(); j++) {
+                auto [idx, val] = inverted_lut_[i][j];
+                scores[idx] += v * float(val);
+            }
+        }
+        for (size_t i = 0; i < n_rows_; ++i) {
+            if ((bitset.empty() || !bitset.test(i)) && scores[i] != 0) {
+                heap.push(i, -scores[i]);
+            }
+        }
+    }
+
+    class Cursor {
+     public:
+        Cursor() = default;
+        Cursor(const std::vector<Neighbor<T>>* lut, size_t num_vec, float max_score, float q_value,
+               const BitsetView* bitset)
+            : lut_(lut), num_vec_(num_vec), max_score_(max_score), q_value_(q_value), bitset_(bitset) {
+            while (loc_ < lut_->size() && !bitset_->empty() && bitset_->test(cur_vec_id())) {
+                loc_++;
+            }
+        }
+        Cursor(const Cursor& rhs) = delete;
+        Cursor(Cursor&& rhs) noexcept {
```

Comment:
Won't `Cursor(Cursor&& rhs) noexcept = default;` work?
Same question for `Cursor& operator=(Cursor&& rhs) noexcept`